packages/jetbrains-plugin/src/main/kotlin/com/mongodb/jbplugin/ActivatePluginPostStartupActivity.kt

```diff
@@ -15,7 +16,7 @@ import kotlinx.coroutines.launch
 class ActivatePluginPostStartupActivity(private val cs: CoroutineScope) : StartupActivity, DumbAware {
     override fun runActivity(project: Project) {
         cs.launch {
-            val pluginActivated = project.getService(PluginActivatedProbe::class.java)
+            val pluginActivated = ApplicationManager.getApplication().getService(PluginActivatedProbe::class.java)
```

Comment:
Is there some documentation where I could learn about the difference between the two/why `ApplicationManager` is preferred for accessing services? Naïvely I would assume that `ApplicationManager.getApplication()` returns a singleton, which admittedly doesn't make a ton of sense to me in combination with dependency/service injection mechanisms

Reply:
There is no much documentation, to be fair, it's only here: https://plugins.jetbrains.com/docs/intellij/plugin-services.html

But basically, you are right in that ApplicationManager.getApplication is a singleton (potentially). The main reason to use the injection through an application or a project it is that if the service is a light service, the actual implementation of the service can change in runtime. An actual real example is when upgrading the plugin, that can be done without restarting the IDE.

The difference between an Application service and a Project service is the scope of:

* The coroutines
* The service itself

In an Application service, the scope of the coroutine is shared across projects. If you have 3 java projects open, and you use the Application coroutine scope, the underlying thread pool can be shared across all projects, reducing the number of threads needed.

In a Project service, you will have a coroutine scope for each project and also a service instance for each project. In some cases this might be desirable (for example, when you are parsing code), but for more generic use cases like telemetry, connecting to MongoDB, and so on is better to just have one single instance.

Reply:
(just going to leave this unresolved because I think it's helpful)